packages/web_benchmarks/lib/client.dart

```diff
@@ -0,0 +1,412 @@
+      sendData: json.encode(_benchmarks.keys.toList()),
+    );
+
+    // 404 is expected in the following cases:
+    // - The benchmark is ran using plain `flutter run`, which does not provide "next-benchmark" handler.
+    // - We ran all benchmarks and the benchmark is telling us there are no more benchmarks to run.
+    if (request.status == 404) {
+      isInManualMode = true;
+      return kManualFallback;
+    }
+
+    isInManualMode = false;
+    return request.responseText;
+  }
+
+  void _checkNotManualMode() {
+    if (isInManualMode) {
+      throw StateError('Operation not supported in manual fallback mode.');
+    }
+  }
+
+  /// Asks the local server to begin tracing performance.
+  ///
+  /// This uses the chrome://tracing tracer, which is not available from within
+  /// the page itself, and therefore must be controlled from outside using the
+  /// DevTools Protocol.
+  Future<void> startPerformanceTracing(String benchmarkName) async {
+    _checkNotManualMode();
+    await html.HttpRequest.request(
+      '/start-performance-tracing?label=$benchmarkName',
+      method: 'POST',
+      mimeType: 'application/json',
+    );
+  }
+
+  /// Stops the performance tracing session started by [startPerformanceTracing].
+  Future<void> stopPerformanceTracing() async {
+    _checkNotManualMode();
+    await html.HttpRequest.request(
+      '/stop-performance-tracing',
+      method: 'POST',
+      mimeType: 'application/json',
+    );
+  }
+
+  /// Sends the profile data collected by the benchmark to the local benchmark
+  /// server.
+  Future<void> sendProfileData(Profile profile) async {
+    _checkNotManualMode();
+    final html.HttpRequest request = await html.HttpRequest.request(
+      '/profile-data',
+      method: 'POST',
+      mimeType: 'application/json',
+      sendData: json.encode(profile.toJson()),
+    );
+    if (request.status != 200) {
+      throw Exception('Failed to report profile data to benchmark server. '
+          'The server responded with status code ${request.status}.');
+    }
+  }
+
+  /// Reports an error to the benchmark server.
+  ///
+  /// The server will halt the devicelab task and log the error.
+  Future<void> reportError(dynamic error, StackTrace stackTrace) async {
+    _checkNotManualMode();
+    await html.HttpRequest.request(
+      '/on-error',
+      method: 'POST',
+      mimeType: 'application/json',
+      sendData: json.encode(<String, dynamic>{
+        'error': '$error',
+        'stackTrace': '$stackTrace',
+      }),
+    );
+  }
+
+  /// Reports a message about the demo to the benchmark server.
+  Future<void> printToConsole(String report) async {
+    _checkNotManualMode();
+    await html.HttpRequest.request(
+      '/print-to-console',
+      method: 'POST',
+      mimeType: 'text/plain',
+      sendData: report,
+    );
+  }
+
+  /// This is the same as calling [html.HttpRequest.request] but it doesn't
+  /// crash on 404, which we use to detect `flutter run`.
```

Comment:
I was going to suggest something like this:
```dart
try {
  return html.HttpRequest.request(url, ...);
} on html.ProgressEvent catch(e) {
  if (xhr.status! == 404) {
    return xhr;
  }
  rethrow;
}
```
but then realized inside the `catch` there's no way to get the `xhr` object :(

Reply:
I cleaned up this method. It's only used once and a lot of cases it's trying to handle don't exist.